### PR TITLE
feat: add github auth 

### LINKED ISSUE #1852: feat: add GitHub Auth using Firebase & display minor info on dashboard
1. Add firebase auth with GitHub to make it effective

   - We basically just need a GitHub Auth, as people who want to contribute must have their GitHub, so that they can directly create issues from the dashboard.
   - We will follow proper conventions like middleware.ts based on the logic
   
- We need to use middleware.ts for the logic part.

For reference: 
Screenshot 2023-10-05 100017
2. We will include a header, with a simple popup to show the data and the basic profile info.

### PR DESCRIPTION
**Pull request by Vidip-Ghosh**
## Fixes Issue

Closes #1852 

## Changes proposed

I've added github firebase auth. Currently, authorization part is in progress. 

## Note to reviewers

I had pushed the changes since I work on gitpod (cloud environment) due to system installation issues & I was fearing loss of changes done by me. As soon as, the work fully gets completed, I will add screen recording.

### CODE REVIEW

**Comment by CBID2** [comment]
@rupali-codes, there seems to be deployment issues.

**Comment by Vidip-Ghosh** [comment]
https://github.com/rupali-codes/LinksHub/assets/91741581/626b114b-7d24-4d43-886d-4925e8aecf1b

Working Video.

**Comment by Vidip-Ghosh** [comment]
https://github.com/rupali-codes/LinksHub/assets/91741581/5b4d40f5-badc-4b46-be8a-095e631f7c2c

Working video of the changes done. 
Now, user can access the resources but can't upvote since he/she isn't authenticated.

**Comment by CBID2** [comment]
> 
> 
> https://github.com/rupali-codes/LinksHub/assets/91741581/5b4d40f5-badc-4b46-be8a-095e631f7c2c
> 
> 
> 
> Working video of the changes done. 
> 
> Now, user can access the resources but can't upvote since he/she isn't authenticated. 

Great progress @Vidip-Ghosh! 😁 I wonder how users can get authenticated? 🤔

**Review by aftabrehan — COMMENTED** [review]
@Vidip-Ghosh, overall looks good but aren't you using the "prettier" formator? The code is hard to read.

**Comment by Vidip-Ghosh** [comment]
> > 658a854e-bd82-482b-a85f-bd83e4382703.mp4
> >     
> >   
> > 
> >   
> > 
> >   
> > 
> > 
> > Working video of the changes done.
> > Now, user can access the resources but can't upvote since he/she isn't authenticated.
> 
> Great progress @Vidip-Ghosh! 😁 I wonder how users can get authenticated? 🤔

Actually, whenever user logs in to `LinksHub`, he/she gets assigned by a session cookie. 
If a user isn't logged in, he/she gets an empty cookie.

**Comment by Vidip-Ghosh** [comment]
> @Vidip-Ghosh, overall looks good but aren't you using the "prettier" formator? The code is hard to read.

Yeah, I haven't used prettier.

**Comment by Vidip-Ghosh** [comment]
https://github.com/rupali-codes/LinksHub/assets/91741581/e74b62ae-8c17-420b-aab5-88a26e8e8102

Adding toast notifications.
Sorry for that firebase runtime error as my firestore client access got expired.

**Comment by Vidip-Ghosh** [comment]
https://github.com/rupali-codes/LinksHub/assets/91741581/ed64f3b7-6ac8-4f69-a85d-37b0a5451839

After making changes.

**Comment by Vidip-Ghosh** [comment]
https://github.com/rupali-codes/LinksHub/assets/91741581/7474d766-44e5-44ac-b8a8-071d1e17db60

Now fixed double occurence of error message.

File: app/SignInWithGithub/SignInWithGithub.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    const router = useRouter();
+    const [imageURL, setImageURL] = useState<string | null>(null);
+    const [authenticated,setAuthenticated] = useState(false);
+    const signIn = async () => {
+        try {
+            const result = await signInWithPopup(auth, provider);
+            const credential = GithubAuthProvider.credentialFromResult(result);
+            console.log("Credential",credential);
+            console.log("Result: ",result);
+            const token: any = result.user.accessToken;
```
Comment (CBID2) [inline app/SignInWithGithub/SignInWithGithub.tsx:20]: @Anmol-Baranwal, according to my terminal, "Property `accessToken' does not exist on type 'User'", so maybe this needs some changing?

File: app/SignInWithGithub/SignInWithGithub.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            const username:any = result.user.displayName;
+            localStorage.setItem('accessToken', token);
+            const currDate = new Date().getTime;
+            document.cookie = `accessToken=${token};path=/; expires=${currDate}`;
+            const imgURL:any = result.user.photoURL;
+            localStorage.setItem('imageURL',imgURL);
```
Comment (CBID2) [inline app/SignInWithGithub/SignInWithGithub.tsx:26]: @Anmol-Baranwal, on this line, my terminal states this:
"Unexpected any. Specify a different type". Maybe this plays a role in the deployment failure?

File: app/SignInWithGithub/SignInWithGithub.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            setImageURL(imgURL);
+            setAuthenticated(true);
```
Comment (CBID2) [inline app/SignInWithGithub/SignInWithGithub.tsx:28]: @Anmol-Baranwal, it also says, "Unexpected any. Specify a different type."

File: components/Cards/Card.tsx [hunk 4]
```diff
@@ -1,20 +1,135 @@
-import { FC, useState, useRef, useEffect } from 'react'
+import React, { FC, useState, useRef, useEffect } from 'react'
 import {BsYoutube , BsPen} from 'react-icons/bs'
 import {AiOutlineRead} from 'react-icons/ai'
 import{MdArticle} from 'react-icons/md'
 import { CopyToClipboard } from 'components/CopyToClipboard/CopyToClipboard'
 import Share from 'components/Share/Share'
 import type { IData } from 'types'
+import { collection, doc,where,query,getDocs, setDoc,getDoc, updateDoc } from 'firebase/firestore'
```
Comment (CBID2) [inline components/Cards/Card.tsx:8]: @Anmol-Baranwal, on my terminal, it says 'updateDoc is defined by not used'

File: components/Cards/Card.tsx [hunk 5]
```diff
@@ -1,20 +1,135 @@
...
+      }
+  
+      const assetDocSnapshot = assetQuerySnapshot.docs[0];
+      const assetDocRef = doc(db, 'resources', data.name);
+      const assetData = assetDocSnapshot.data();
+      const upvotes = assetData.upvotes || {};
+      const userUid = auth.currentUser? auth.currentUser.uid : null;
+      console.log("User ID: ",userUid)
+  
+      if (upvotes[userUid]) {
```
Comment (CBID2) [inline components/Cards/Card.tsx:77]: @Anmol-Baranwal, according to my terminal, "Type null can't be used as an index type"(this message appears for lines 79 and 82 as well). Could this also play role in the deployment failure?

**Comment by Anmol-Baranwal** [comment]
@CBID2 
Yeah, these are the errors, and I'll work on fixing them. I use VSCode, so it deploys even if there are linting errors and warnings.

The same applies to the other PR.

Leave it to us because you need to set up Firebase auth and the database with .env, then you can check it properly.

If you're interested in doing it, let me know, and I'll share the resource for which you can set it up. It's your choice, but it does require considerable effort, so I wouldn't recommend it.

**Comment by CBID2** [comment]
> @CBID2 
> 
> Yeah, these are the errors, and I'll work on fixing them. I use VSCode, so it deploys even if there are linting errors and warnings.
> 
> 
> 
> The same applies to the other PR.
> 
> 
> 
> Leave it to us because you need to set up Firebase auth and the database with .env, then you can check it properly.
> 
> 
> 
> If you're interested in doing it, let me know, and I'll share the resource for which you can set it up. It's your choice, but it does require considerable effort, so I wouldn't recommend it.

Great!

**Comment by Anmol-Baranwal** [comment]
> Great!

Do you want to do it or not? Let me know. 😅

**Comment by rupali-codes** [comment]
@Vidip-Ghosh @CBID2 @Anmol-Baranwal it's saying this
image

**Comment by Anmol-Baranwal** [comment]
I will check it by tomorrow. Maybe he used firebase sdk.

**Comment by Vidip-Ghosh** [comment]
Screenshot 2023-12-13 at 16 15 54
Actually that accessToken is coming from here. As soon as user signs in. 
@rupali-codes 

That's why I used `result.user.accessToken`

**Review by Anmol-Baranwal — COMMENTED** [review]
@rupali-codes 

image

I don't think we can test it on localhost, even if it is a trusted domain in firebase.

I tried forwarding port using tunnel in vs code and even ngrok for callback url.

One hell of a task :)

image

Can you check at your side? 

It deployed for once giving promise unresolved error, then I restarted the server and there is a problem with auth domain.

**Comment by Anmol-Baranwal** [comment]
@Vidip-Ghosh 
First fix the errors and warnings. `npm run lint`

image

Then, it will deploy for sure.

**Comment by Vidip-Ghosh** [comment]
> @Vidip-Ghosh First fix the errors and warnings. `npm run lint`
> 
> image
> 
> Then, it will deploy for sure.

Okay alright

**Comment by Vidip-Ghosh** [comment]
Screenshot 2023-12-19 at 10 15 42
@Anmol-Baranwal although I have fixed the warnings on my end but on running `npm run lint` these things are coming. Is this the reason of deployment getting failed still?

**Comment by rupali-codes** [comment]
@Vidip-Ghosh & @Anmol-Baranwal 

image

**Review by CBID2 — APPROVED** [review]
It looks good now @Anmol-Baranwal. @rupali-codes seemed to fix the issues we pointed out.

**Comment by Vidip-Ghosh** [comment]
I had a discussion with @rupali-codes, the changes of `Auth` part will be merged into `firebase-auth` branch. So therefore I had closed this one & opened PR #2213 to get merged into `firebase-auth` branch.

**Pull request closed without merging**